src/libsyntax/ast.rs

```diff
@@ -368,8 +393,8 @@ pub enum WherePredicate {
 #[derive(Clone, PartialEq, Eq, RustcEncodable, RustcDecodable, Hash, Debug)]
 pub struct WhereBoundPredicate {
     pub span: Span,
-    /// Any lifetimes from a `for` binding
-    pub bound_lifetimes: Vec<LifetimeDef>,
+    /// Any generics from a `for` binding
+    pub bound_generics: Generics,
```

Comment:
There may be reasons to change `Vec<LifetimeDef>` into `Vec<GenericParam>` in the future, but why `Generics`?
I'd rather just leave `Vec<LifetimeDef>` in place, this change doesn't help const generics in any way and permits too many things in AST that can't happen in actual code.

Reply:
Same for `BareFnTy` and `PolyTraitRef`.

Reply:
IIRC there are a bunch of places where `visit_lifetime_def` is implemented separately from type parameters (`visit_generics`), only because `visit_generics` wouldn't be called in those `Vec<LifetimeDef>` cases (see also https://github.com/rust-lang/rust/pull/45930#discussion_r150382826).

Reply:
I'm not sure the current situation is significantly worse than a bunch of impossible "`span_bug`" code.

If `visit_lifetime_def` is removed and `visit_generic_param` is uniformly used for visiting (https://github.com/rust-lang/rust/pull/45930#discussion_r150382826) then `Generics` here could be replaced with `Vec<GenericParam>` improving situation somewhat.

Reply:
FWIW I expect @nikomatsakis would eventually want HRTB to include `where` clauses too, on top of `for<T: Trait>`, but I don't know the exact plans.
`Vec<GenericParam>` is fine with me too.